### ISSUE TITLE
Better Stop in `MessageReceiverNotifer`

### ISSUE THREAD

**Issue by tmasternak** [Particular/NServiceBus.AzureServiceBus#439]
## Context
When stopping `MessagePump` we execute `Stop` methods for all `MessageReceiverNotifer`es. The [logic](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/develop/src/Transport/Receiving/MessageReceiverNotifier.cs#L98) of `Stop` consists of couple of steps:
 1. Set `stopping` flag //after that all messages pushed by SDK to [ProcessMessages](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/develop/src/Transport/Receiving/MessageReceiverNotifier.cs#L180) are ignored
 2. Wait on all in-flight messages stored in `pipelineInvocationTasks` with 30s of timeout
 3. Shutdown internal receivers //after that no more messages are pushed by SDK

Let's analyze what is happening in different receive modes ie. `PeekLock` and `ReceiveAndDelete`:
 * With `PeekLock` after we are done with `1` we will wait for in-flight messages to let them finish correctly (active receivers are needed to `Complete`/`Abandon` `BrokeredMessage`) and close receives. As a result during the shutdown time we might ignore some messages pushed by SDK but because we are in `PeekLock` they will reappear in the input queue after lock period elapses.
 * With `RecieveAndDelete` after we are done with `1` we will keep ignoring messages until we are done with in-flight messages. This is worse than in case of `PeekLock` as we will not get those messages back. What is more in this mode we do not need active receivers because we do don't have to `Complete`\`Abandon` operations.

In summary with current approach we are not breaking any guarantees - in `ReceiveAndDelete` one can lose messages but I think we could do better. 

### Possible solutions
I don't have a ready solution for this but here are some ideas:
1. Change the order of operations in `Stop`
2. Have two separate approaches for `Stop`ing depending on the mode
3. Follow the ASQ approach /cc: @Scooletz 

Connects to https://github.com/Particular/NServiceBus.AzureServiceBus/issues/587

**Comment by Scooletz** [comment]
The ASQ approach. Let's introduce two separate strategies for different guarantees.

**Comment by SeanFeldman** [comment]
I'll be the 👿 advocate. We discourage people from running in `ReceiveAndDelete` mode. By default, we're using `PeekLock`. Those that opt into `ReceiveAndDelete` do that knowing that they will loose messages.

Now with that on the table, I do like the idea of doing better than we do now. What those 2 strategies would look like? Wouldn't be it simpler to have 2 `Stop` approaches? I.e. can't say one or another 😄 Let's go over this in our sync?

**Comment by Scooletz** [comment]
Sure. We can discuss it.

**Comment by danielmarbach** [comment]
We'll address this when we start working on the new ASB client. Closing
